### ISSUE TITLE
the plugin does not work with iRODS 4.2.1 on CENTOS 7

### ISSUE THREAD

**Issue by ccacciari** [irods/irods_rule_engine_plugin_audit_amqp#15]
I have deployed the amqp plugin on a CENTOS 7, using the packages from the iRODS repo: renci-irods/7/x86_64. The following packages are installed on the system:
```
irods-externals-avro1.7.7-0-1.0-1.x86_64
irods-externals-libarchive3.1.2-0-1.0-1.x86_64
irods-icommands-4.2.1-1.x86_64
irods-microservice-plugins-curl-2.1.0-1.x86_64
irods-externals-cmake3.5.2-0-1.0-1.x86_64
irods-externals-autoconf5ad3567c-0-1.0-1.x86_64
irods-externals-epm4.2-0-1.0-1.x86_64
irods-externals-boost1.60.0-0-1.0-1.x86_64
irods-externals-zeromq4-14.1.3-0-1.0-1.x86_64
irods-externals-jansson2.7-0-1.0-1.x86_64
irods-runtime-4.2.1-1.x86_64
irods-server-4.2.1-1.x86_64
irods-auth-plugin-gsi-2.1.0-1.x86_64
irods-externals-qpid-with-proton0.34-0-1.0-1.x86_64
irods-externals-libs3a30e55e8-0-1.0-1.x86_64
irods-externals-cppzmq4.1-0-1.0-1.x86_64
irods-externals-clang3.8-0-1.0-1.x86_64
irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64
irods-externals-clang-runtime3.8-0-1.0-1.x86_64
irods-database-plugin-postgres-4.2.1-1.x86_64
irods-devel-4.2.1-1.x86_64
irods-externals-imagemagick7.0.3-0-1.0-1.x86_64
irods-externals-libs3f6a2098e-0-1.0-1.x86_64
```
iRODS works fine, but when I configure the amqp plugin in the server_config.json, it fails with the following error:
```
Jul 28 16:22:36 pid:22959 remote addresses: 127.0.0.1, ::1 ERROR: applyRule: -1822000, [-]      /home/irodsbuild/irods/server/re/include/irods_re_plugin.hpp:321:irods::error irods::dynamic_operation_execution_manager<std::__1::tuple<>, RuleExecInfo *, irods::rule_execution_manager_pack::DONT_AUDIT_RULE>::call(std::string, std::string, OP, As &&...) [T = std::__1::tuple<>, C = RuleExecInfo *, Audit = irods::rule_execution_manager_pack::DONT_AUDIT_RULE, OP = std::__1::function<irods::error (const std::__1::basic_string<char> &, irods::re_pack_inp<std::__1::tuple<> > &)>, As = <const std::__1::basic_string<char> &, irods::re_pack_inp<std::__1::tuple<> > &>] :  status [INVALID_ANY_CAST]  errno [] -- message []
        [-]     /home/irodsbuild/irods/server/re/include/irods_re_plugin.hpp:201:irods::error irods::pluggable_rule_engine<std::__1::tuple<> >::rule_exists(const std::string &, T &, bool &) [T = std::__1::tuple<>] :  status [INVALID_ANY_CAST]  errno [] -- message [failed to extract rule_exists operation from instance [irods_rule_engine_plugin-audit_amqp-instance] exception message [boost::bad_any_cast: failed conversion using boost::any_cast]]

Jul 28 16:22:36 pid:22959 remote addresses: 127.0.0.1, ::1 ERROR: applyRule: -1822000, [-]      /home/irodsbuild/irods/server/re/include/irods_re_plugin.hpp:201:irods::error irods::pluggable_rule_engine<std::__1::tuple<> >::rule_exists(const std::string &, T &, bool &) [T = std::__1::tuple<>] :  status [INVALID_ANY_CAST]  errno [] -- message [failed to extract rule_exists operation from instance [irods_rule_engine_plugin-audit_amqp-instance] exception message [boost::bad_any_cast: failed conversion using boost::any_cast]]

Jul 28 16:22:36 pid:22959 remote addresses: 127.0.0.1, ::1 ERROR: initAgent: acChkHostAccessControl error, status = -1822000
Jul 28 16:22:36 pid:22959 remote addresses: 127.0.0.1, ::1 ERROR: agentMain :: initAgent failed: -1822000
Jul 28 16:22:36 pid:22906  ERROR: Agent process [22959] exited with status [1]
```
this is the configuration I set in the server_config.json:
 ```
       "rule_engines": [
            {
                "instance_name": "irods_rule_engine_plugin-irods_rule_language-instance", 
                "plugin_name": "irods_rule_engine_plugin-irods_rule_language", 
                "plugin_specific_configuration": {
                    "re_data_variable_mapping_set": [
                        "core"
                    ], 
                    "re_function_name_mapping_set": [
                        "core"
                    ], 
                    "re_rulebase_set":["core"], 
                    "regexes_for_supported_peps": [
                        "ac[^ ]*", 
                        "msi[^ ]*", 
                        "[^ ]*pep_[^ ]*_(pre|post)"
                    ]
                }, 
                "shared_memory_instance": "irods_rule_language_rule_engine"
            }, 
            {
                "instance_name": "irods_rule_engine_plugin-cpp_default_policy-instance", 
                "plugin_name": "irods_rule_engine_plugin-cpp_default_policy", 
                "plugin_specific_configuration": {}
            },
            {
                "instance_name": "irods_rule_engine_plugin-audit_amqp-instance",
                "plugin_name": "irods_rule_engine_plugin-audit_amqp",
                "plugin_specific_configuration" : {
                    "amqp_location" : "ANONYMOUS@localhost:5672",
                    "amqp_options" : "",
                    "amqp_topic" : "audit_messages",
                    "pep_regex_to_match" : "audit_.*"
                }
            }
        ]
    }, 
    "rule_engine_namespaces": [
        "",
        "audit_",
        "indexing_"
    ],
```

**Comment by trel** [comment]
We are working on getting a released package out the door ASAP.  The current master branch works as expected, and we are working on the test framework to get it into CI.

If you can build the current master branch, you should be able to install it and avoid the `INVALID_ANY_CAST` errors.

You may also want to have `irods_rule_engine_plugin-audit_amqp-instance` listed before `irods_rule_engine_plugin-cpp_default_policy-instance `.

**Comment by ccacciari** [comment]
Thanks for the advises. 
However if I try to deploy with yum the package built from the source of the current master branch, I got:
```
$ sudo yum install /home/irodsmaster/irods_rule_engine_plugin_audit_amqp/build/irods-rule-engine-plugin-audit-amqp-4.2.1-centos7-x86_64.rpm
Loaded plugins: fastestmirror
Examining /home/irodsmaster/irods_rule_engine_plugin_audit_amqp/build/irods-rule-engine-plugin-audit-amqp-4.2.1-centos7-x86_64.rpm: irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64
Marking /home/irodsmaster/irods_rule_engine_plugin_audit_amqp/build/irods-rule-engine-plugin-audit-amqp-4.2.1-centos7-x86_64.rpm to be installed
Resolving Dependencies
--> Running transaction check
---> Package irods-rule-engine-plugin-audit-amqp.x86_64 0:4.2.1-1 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

===================================================================================================================
 Package                            Arch   Version Repository                                                 Size
===================================================================================================================
Installing:
 irods-rule-engine-plugin-audit-amqp
                                    x86_64 4.2.1-1 /irods-rule-engine-plugin-audit-amqp-4.2.1-centos7-x86_64 498 k

Transaction Summary
===================================================================================================================
Install  1 Package

Total size: 498 k
Installed size: 498 k
Is this ok [y/d/N]: y
Downloading packages:
Running transaction check
Running transaction test

Transaction check error:
  file /var from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package filesystem-3.2-21.el7.x86_64
  file /var/lib from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package filesystem-3.2-21.el7.x86_64
  file /usr/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-runtime-4.2.1-1.x86_64
  file /usr/lib/irods/plugins from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-runtime-4.2.1-1.x86_64
  file /var/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-icommands-4.2.1-1.x86_64
  file /var from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-icommands-4.2.1-1.x86_64
  file /var/lib from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-icommands-4.2.1-1.x86_64
  file /usr/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-server-4.2.1-1.x86_64
  file /var/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-server-4.2.1-1.x86_64
  file /var/lib/irods/scripts/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-server-4.2.1-1.x86_64
  file /usr/lib/irods/plugins from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-server-4.2.1-1.x86_64
  file /usr/lib/irods/plugins/rule_engines from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-server-4.2.1-1.x86_64
  file /var from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-server-4.2.1-1.x86_64
  file /var/lib from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-server-4.2.1-1.x86_64
  file /var/lib/irods/scripts from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-server-4.2.1-1.x86_64
  file /var/lib/irods/scripts/irods/test from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-server-4.2.1-1.x86_64
  file /usr/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-database-plugin-postgres-4.2.1-1.x86_64
  file /var/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-database-plugin-postgres-4.2.1-1.x86_64
  file /usr/lib/irods/plugins from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-database-plugin-postgres-4.2.1-1.x86_64
  file /var from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-database-plugin-postgres-4.2.1-1.x86_64
  file /var/lib from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-database-plugin-postgres-4.2.1-1.x86_64
  file /usr/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-auth-plugin-gsi-2.1.0-1.x86_64
  file /var/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-auth-plugin-gsi-2.1.0-1.x86_64
  file /var/lib/irods/scripts/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-auth-plugin-gsi-2.1.0-1.x86_64
  file /usr/lib/irods/plugins from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-auth-plugin-gsi-2.1.0-1.x86_64
  file /var from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-auth-plugin-gsi-2.1.0-1.x86_64
  file /var/lib from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-auth-plugin-gsi-2.1.0-1.x86_64
  file /var/lib/irods/scripts from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-auth-plugin-gsi-2.1.0-1.x86_64
  file /var/lib/irods/scripts/irods/test from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-auth-plugin-gsi-2.1.0-1.x86_64
  file /usr/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-microservice-plugins-curl-2.1.0-1.x86_64
  file /var/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-microservice-plugins-curl-2.1.0-1.x86_64
  file /var/lib/irods/scripts/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-microservice-plugins-curl-2.1.0-1.x86_64
  file /usr/lib/irods/plugins from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-microservice-plugins-curl-2.1.0-1.x86_64
  file /var from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-microservice-plugins-curl-2.1.0-1.x86_64
  file /var/lib from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-microservice-plugins-curl-2.1.0-1.x86_64
  file /var/lib/irods/scripts from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-microservice-plugins-curl-2.1.0-1.x86_64
  file /var/lib/irods/scripts/irods/test from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-microservice-plugins-curl-2.1.0-1.x86_64
  file /usr/lib/irods from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package irods-devel-4.2.1-1.x86_64
  file /var from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package davrods-4.2.1_1.1.1-1.x86_64
  file /var/lib from install of irods-rule-engine-plugin-audit-amqp-4.2.1-1.x86_64 conflicts with file from package davrods-4.2.1_1.1.1-1.x86_64
```

**Comment by trel** [comment]
Curious.  That shouldn't be happening.

We'll try to reproduce that here, but we're not seeing it at the moment.

**Comment by trel** [comment]
We have this building in CI now - and have closed #16.

This plugin will be released alongside iRODS 4.2.2 within the next couple days.
